File: sorting.py
def bubbleSort(arr):
    for _ in arr:
        for i in range(len(arr) - 1):
            if (arr[i+1] < arr[i]):
                swap(i, i+1, arr)
            else: 
                continue
    return arr


def selectionSort(arr):
    for i in range(len(arr)):
        minIdx = i
        minChanged = False
        for j in range(i+1,len(arr)):
            if arr[j] < arr[minIdx]:
                minIdx = j
                minChanged = True
        if minChanged:
            swap(i, minIdx, arr)
    return arr


def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j+1] = arr[j]
            j -= 1
        arr[j+1] = key
    return arr


def mergeSort(arr):
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]
    return merge(mergeSort(left), mergeSort(right))

def merge(left, right):
    leftIdx, rightIdx = 0, 0
    res = []
    while leftIdx < len(left) and rightIdx < len(right):
        if (left[leftIdx] < right[rightIdx]):
            res.append(left[leftIdx])
            leftIdx += 1
        else:
            res.append(right[rightIdx])
            rightIdx += 1
    return res + left[leftIdx:] + right[rightIdx:]


def quickSort(arr):
    if len(arr) <= 1:
        return arr
    quick(arr, 0, len(arr)-1)
    return arr

def quick(arr, start, end):
    if start >= end:
        return
    pivot = start
    left = start + 1
    right = end

    while left <= right:
        if (arr[left] >= arr[pivot] and arr[right] <= arr[pivot]):
            swap(left, right, arr)
        if (arr[left] <= arr[pivot]):
            left += 1
        if (arr[right] >= arr[pivot]):
            right -= 1
    swap(pivot, right, arr)
    quick(arr, start, right-1)
    quick(arr, right+1, end)
    return


def heapify(arr, n, i):
    largest = i
    left_child = 2 * i + 1
    right_child = 2 * i + 2

    if left_child < n and arr[left_child] > arr[largest]:
        largest = left_child

    if right_child < n and arr[right_child] > arr[largest]:
        largest = right_child

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

def heap_sort(arr):
    n = len(arr)

    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  
        heapify(arr, i, 0)
    return arr


def swap(i, j, arr):
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp
    return arr

test = [5, 1, 0, 2, 8, 4, 6, 1]
print(insertion_sort(test))

File: sorting.js
function bubbleSort(arr) {
    for (let _ of arr) {
        for (let j = 0; j < arr.length - 1; j++) {
            if (arr[j] > arr[j+1]) {
                swap(j, j+1, arr);
            } else {
                continue;
            }
        }
    }
    return arr
}


function selectionSort(arr) {
    for (let i = 0; i < arr.length; i++) {
        let min = i;
        let minChanged = false;
        for (let j = i+1; j < arr.length; j++) {
            if (arr[j] < arr[min]) {
                min = j;
                minChanged = true;
            }
        }
        if (minChanged) {
            swap(min, i, arr);
        }
    }
    return arr;
}


function insertionSort(arr) {
    for (let i = 0; i < arr.length; i++) {
        if (arr[i] < arr[0]) {
            arr.unshift(arr.splice(i, 1)[0]);
        } else {
            for (let j = 1; j < arr.length; j++) {
                if (arr[i] <= arr[j] && arr[i] >= arr[j-1]) {
                    arr.splice(j, 0, arr.splice(i, 1)[0]);
                }
            }
        }
    }
    return arr;
}


function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }
    let mid = arr.length / 2;
    let left = arr.slice(0,mid);
    let right = arr.slice(mid);
    return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
    let res = [];
    let leftIdx = 0;
    let rightIdx = 0;
    while (leftIdx < left.length && rightIdx < right.length) {
        if (left[leftIdx] < right[rightIdx]) {
            res.push(left[leftIdx]);
            leftIdx++;
        } else {
            res.push(right[rightIdx]);
            rightIdx++;
        }
    }
    return res.concat(left.slice(leftIdx)).concat(right.slice(rightIdx));
}


function quickSort(arr) {
    if (arr.length <= 1) return arr;
    quickHelper(arr, 0, arr.length-1);
    return arr;
}

function quickHelper(arr, start, end) {
    if (start >= end) return;
    let pivot = start;
    let left = start + 1;
    let right = end;
    
    while (left <= right) {
        if (arr[left] > arr[pivot] && arr[right] < arr[pivot]) {
            swap(left, right, arr);
        }
        if (arr[left] <= arr[pivot]) {
            left++;
        }
        if (arr[right] >= arr[pivot]) {
            right--;
        }
    }
    swap(pivot, right, arr);
    quickHelper(arr, start, right-1);
    quickHelper(arr, right+1, end);
    return 
}

function swap(i, j, arr) {
    let temp = arr[i];    
    arr[i] = arr[j];
    arr[j] = temp;
    return arr;
}


const test = [3, 1, 5, 2, 0, 7, 2, 10];
console.log(quickSort(test));